### PR TITLE
[SPARK-35616][PYTHON] Make `astype` method data-type-based

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?

Make `astype` method data-type-based.

**Non-goal: Match pandas' `astype` TypeErrors.**
Currently, `astype` throws TypeError error messages only when the destination type is not recognized. However, for some destination types that don't make sense to the specific type of  Series/Index, for example, `numeric Series/Index → bytes`, we don't have proper TypeError error messages.
Since the goal of the PR is refactoring mainly, the above issue might be resolved later if needed.

### Why are the changes needed?

There are many type checks in the `astype` method. Since `DataTypeOps` and its subclasses are introduced, we should refactor `astype` to make it data-type-based. In this way, code is cleaner, more maintainable, and more flexible.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit tests.

Keyword: SPARK-35337

### CODE REVIEW

**Comment by xinrong-databricks** [comment]
CC @ueshin @HyukjinKwon @itholic

File: python/pyspark/pandas/data_type_ops/binary_ops.py [hunk 1]
```diff
@@ -53,3 +57,34 @@ def radd(self, left, right) -> Union["Series", "Index"]:
             raise TypeError(
                 "Concatenation can not be applied to %s and the given type." % self.pretty_name
             )
+
+    def astype(
+        self, index_ops: Union["Index", "Series"], dtype: Union[str, type, Dtype]
+    ) -> Union["Index", "Series"]:
+        dtype, spark_type = pandas_on_spark_type(dtype)
+        if not spark_type:
+            raise ValueError("Type {} not understood".format(dtype))
```
Comment (ueshin) [inline python/pyspark/pandas/data_type_ops/binary_ops.py:66]: Does this happen?

**Review by ueshin — COMMENTED** [review]
Seems like we can extract more methods and reuse. e.g., most of `if isinstance(spark_type, BooleanType):` or `elif isinstance(spark_type, StringType):` clauses look same.
Could you try?

File: python/pyspark/pandas/data_type_ops/binary_ops.py (hunk 1 above)
Comment (xinrong-databricks) [reply python/pyspark/pandas/data_type_ops/binary_ops.py]: Good catch, updated.

File: python/pyspark/pandas/tests/data_type_ops/test_binary_ops.py [hunk 2]
```diff
@@ -147,6 +147,14 @@ def test_from_to_pandas(self):
         self.assert_eq(pser, psser.to_pandas())
         self.assert_eq(ps.from_pandas(pser), psser)
 
+    def test_astype(self):
+        pser = self.pser
+        psser = self.psser
+        self.assert_eq(pd.Series(["1", "2", "3"]), psser.astype(str))
+        self.assert_eq(pser.astype("category"), psser.astype("category"))
+        cat_type = pd.api.types.CategoricalDtype(categories=[b"2", b"3", b"1"])
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_binary_ops.py:155]: nit: Shall we import `CategoricalDtype` at the header? And ditto in the other files.

File: python/pyspark/pandas/data_type_ops/base.py [hunk 3]
```diff
@@ -39,9 +41,8 @@
     TimestampType,
     UserDefinedType,
 )
-
 import pyspark.sql.types as types
```
Comment (ueshin) [inline python/pyspark/pandas/data_type_ops/base.py:44]: I'm wondering why we import some types from `pyspark.sql.types` but also import `types` here?

Comment (xinrong-databricks) [reply python/pyspark/pandas/data_type_ops/base.py]: Makes sense, adjusted.

File: python/pyspark/pandas/tests/data_type_ops/test_binary_ops.py (hunk 2 above)
Comment (xinrong-databricks) [reply python/pyspark/pandas/tests/data_type_ops/test_binary_ops.py]: Certainly.

**Review by ueshin — APPROVED** [review]
LGTM.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**